packages/json-rpc-engine/src/v2/JsonRpcEngineV2.ts

```diff
@@ -0,0 +1,342 @@
+   * @param options - The options for the handle operation.
+   * @param options.context - The context to pass to the middleware.
+   */
+  async handle(
+    notification: Request & JsonRpcNotification,
+    options?: HandleOptions,
+  ): Promise<void>;
+
+  async handle(
+    request: Request,
+    { context }: HandleOptions = {},
+  ): Promise<Result | void> {
+    const isReq = isRequest(request);
+    const { result } = await this.#handle(request, context);
+
+    if (isReq && result === undefined) {
+      throw new JsonRpcEngineError(
+        `Nothing ended request: ${stringify(request)}`,
+      );
+    }
+    return result;
+  }
+
+  /**
+   * Handle a JSON-RPC call (i.e. request or notification).
+   *
+   * This exists because a {@link JsonRpcCall} overload of {@link handle} cannot
+   * coexist with the other overloads of that method. In particular, such an
+   * overload will always return `void`.
+   *
+   * @param request - The JSON-RPC call to handle.
+   * @param options - The options for the handle operation.
+   * @param options.context - The context to pass to the middleware.
+   * @returns The JSON-RPC response, or `void` if the call is a notification.
+   */
+  async handleAny(
+    request: JsonRpcCall & Request,
+    options?: HandleOptions,
+  ): Promise<Result | void> {
+    return this.handle(request, options);
+  }
+
+  /**
+   * Handle a JSON-RPC request. Throws if a middleware performs an invalid
+   * operation. Permits returning an `undefined` result.
+   *
+   * @param originalRequest - The JSON-RPC request to handle.
+   * @param context - The context to pass to the middleware.
+   * @returns The result from the middleware.
+   */
+  async #handle(
+    originalRequest: Request,
+    context: MiddlewareContext = new MiddlewareContext(),
+  ): Promise<{
+    result: Result | void;
+    request: Readonly<Request>;
+  }> {
+    this.#assertIsNotDestroyed();
+
+    deepFreeze(originalRequest);
+
+    const state: RequestState<Request, Result> = {
+      request: originalRequest,
+      result: undefined,
+    };
+    const middlewareIterator = this.#makeMiddlewareIterator();
+    const firstMiddleware = middlewareIterator.next().value;
+
+    const makeNext = this.#makeNextFactory(middlewareIterator, state, context);
+
+    const result = await firstMiddleware({
+      request: originalRequest,
+      context,
+      next: makeNext(),
+    });
+    this.#updateResult(result, state);
+
+    return state;
+  }
+
+  /**
+   * Create a factory of `next()` functions for use with a particular request.
+   * The factory is recursive, and a new `next()` is created for each middleware
+   * invocation.
+   *
+   * @param middlewareIterator - The iterator of middleware for the current
+   * request.
+   * @param state - The current values of the request and result.
+   * @param context - The context to pass to the middleware.
+   * @returns The `next()` function factory.
+   */
+  #makeNextFactory(
```

Comment:
Any thoughts on whether recursively running middlewares will have performance implications versus the current iterative approach?